content/client-lib-development-guide/features.textile

```diff
@@ -74,21 +74,32 @@ h3(#restclient). RestClient
+* @(RSA14)@ If Token Auth is the default authentication scheme, an exception will be raised if a token is not provided or there is no means to generate a token. For example, if only the option @useTokenAuth@ is specified, and a @key@ is not provided, then the client library is unable to authenticate or issue a token
+* @(RSA15)@ If Token Auth is being used and additionally @clientId@ has been set in the @ClientOptions@ when the library was instanced, the @clientId@ of the token must match the @clientId@ provided in @ClientOptions@. Following an auth request which uses a @TokenDetails@ or @TokenRequest@ object that contains an incompatible @clientId@, the library should change the connection state to @FAILED@ and emit an error
 * @(RSA5)@ TTL for new tokens is specified in milliseconds and defaults to the REST API default (1 hour)
 * @(RSA6)@ The @capability@, if not specified, defaults to allow all operations permitted for the key associated with the token
-* @(RSA7)@ If a @clientId@ is provided in the options, then the string will be encoded in UTF-8 format before it is sent to Ably
+* @(RSA7)@ @clientId@ and authenticated clients:
+** @(RAS7a)@ If a @clientId@ is present in the authentication token, or in the case of a realtime connection, the endpoint has an explicit @clientId@ param configured, then the client is considered to be authenticated i.e. it has an identity that is implicit in all operations. The following applies to authenticated clients:
+*** @(RAS7a1)@ All operations (such as message publishing or presence) will have an implicit @clientId@. The Ably service automatically updates the @clientId@ attribute (when empty) for all @Message@ and @PresenceMessage@ messages received from that authenticated client, and any messages then published from the Ably service, will have the @clientId@ attribute populated. It is therefore expected that Ably client libraries do not explicitly set the @clientId@ field on messages published when @clientId@ is implicit in the connection or authentication scheme
+*** @(RSA7a2)@ The server will reject the publishing of messages with an explicit @clientId@ that does not match the authenticated client's @clientId@. The connection and channel remains available for further operations
+*** @(RSA7a3)@ The client library raises an exception if the @clientId@ attribute of any @Message@ or @PresenceMessage@ published is not @null@ and does not match the non-@null@ @Auth#clientId@. The connection and channel remains available for further operations
+*** @(RSA7a4)@ @Auth#clientId@ attribute returns a string value for the authenticated client's @clientId@. If the client is not authenticated and a @clientId@ value was not provided when instanced with @ClientOptions@, then @Auth#clientId@ is @null@
+** @(RSA7e)@ @Auth#clientId@ is not @null@ when:
+*** @(RSA7e1)@ A @clientId@ is provided in the @ClientOptions@. @clientId@ should be a UTF-8 string
```

Comment:
Yes, we've discussed this before, whilst in Java this may not be an issue, I thought we had agreed that @clientId@, like @name@ or channels and events, is always expected to be UTF-8 string.  Therefore, if someone were to provide a string in another format, the client library should encode as UTF-8.  See https://github.com/ably/ably-ruby/blob/master/spec/unit/auth_spec.rb#L13-L67 for a set of unit tests in the Ruby lib